docs/proposals/templates.md

```diff
@@ -0,0 +1,404 @@
+6. Return the processed template object.
+
+The client can now either return the processed template to the user in a desired form (eg json or yaml), or directly iterate the 
+api objects within the template, invoking the appropriate object creation api endpoint for each element.
+
+The result is a consistently recreatable application configuration, including well-defined labels for grouping objects created by 
+the template, with end-user customizations as enabled by the template author.
+
+#### Template Authoring
+
+To aid application authors in the creation of new templates, it should be possible to export existing objects from a project 
+in template form.  A user should be able to export all or a filtered subset of objects from a namespace, wrappered into a 
+Template API object.  The user will still need to customize the resulting object to enable parameterization and labeling, 
+though sophisticated export logic could attempt to auto-parameterize well understood api fields.  Such logic is not considered 
+in this proposal.
+
+#### Tooling
+As described above, templates can be instantiated by posting them to a template processing endpoint.  CLI tools should 
+exist which can input parameter values from the user as part of the template instantiation flow.  
+
+More sophisticated UI implementations should also guide the user through which parameters the template expects, the description 
+of those templates, and the collection of user provided values.
+
+In addition, as described above, existing objects in a namespace can be exported in template form, making it easy to recreate a 
+set of objects in a new namespace or a new cluster.
+
+
+## Examples
+
+#### Example Templates
+
+* [Jenkins template](https://github.com/openshift/origin/blob/master/examples/jenkins/jenkins-persistent-template.json)
+* [MySQL DB service template](https://github.com/openshift/origin/blob/master/examples/db-templates/mysql-persistent-template.json)
+
+#### Examples of OpenShift Parameter Usage
+
+(mapped to use cases described above)
+
+* [Generate and share passwords](https://github.com/jboss-openshift/application-templates/blob/master/eap/eap64-mongodb-s2i.json#L146-L152)
+* [Simple deployment-time customization of “app” configuration via environment values](https://github.com/jboss-openshift/application-templates/blob/master/eap/eap64-mongodb-s2i.json#L108-L126) (eg memory tuning, resource limits, etc)
+* [Customization of component names with referential integrity](https://github.com/jboss-openshift/application-templates/blob/master/eap/eap64-mongodb-s2i.json#L199-L207)
+* [Customize cross-component references](https://github.com/jboss-openshift/application-templates/blob/master/eap/eap64-mongodb-s2i.json#L78-L83) (eg user provides the name of a secret that already exists in their namespace, to use in a pod as a TLS cert)
+
+## Known Limitations
+
+* Generators must be statically compiled into the template api endpoint
```

Comment:
This is a significant limitation. By contrast, DM allows an open ended set of templates to be stored in registries, and referenced on demand. Templates can be written in Python, allowing for open ended generation capability. No changes to the Kubernetes master are required to introduce new capabilities.